### PR TITLE
[A2-872] and [A2-817] Endpoints for apply/cancel project rule updates

### PR DESCRIPTION
**Pull request by msorens**
### :nut_and_bolt: Description

Provide an API call to apply rules and cancel rules.
This is now a combination A2-872 and A2-817.

#### Setup

```
start_all_services
rebuild components/automate-gateway
rebuild components/authz-service
chef-automate iam `upgrade-to-v2 --beta2.1
export TOK=`chef-automate iam token create my-token --admin`
```

#### Insert Nodes

Note that you'll have to add more than 100 for ingest to take long enough to see the status but for starters:

```
# chef_load_nodes 100
```

Then let's see how many nodes for each chef-server org exist (in this case "The Misfits" and everything is in unassigned because we haven't made rules yet):

```
# curl -H "projects: (unassigned)" -sH "api-token: $TOK" https://localhost/api/v0/cfgmgmt/nodes?pagination.size=1000 --insecure | jq . | grep "The Misfits" | wc -l
      9
```

#### Make projects and rules

```
# curl -kH "api-token: $TOK" -XPOST -d "$(jq -n '{ id: "project1", name: "my project 1" }')" https://localhost/apis/iam/v2beta/projects
# curl -kH "api-token: $TOK" -XPOST -d "$(jq -n '{ id: "project2", name: "my project 2" }')" https://localhost/apis/iam/v2beta/projects
```

```
# curl -kH "api-token: $TOK" https://localhost/apis/iam/v2beta/rules -d "$(jq -n '{ id: "rule1", name: "my foo rule", type: "NODE", project_id: "project1", conditions: [{ operator: "EQUALS", attribute: "CHEF_ORGS", values: ["The Misfits"]}]}')" | jq .
# curl -kH "api-token: $TOK" https://localhost/apis/iam/v2beta/rules -d "$(jq -n '{ id: "rule2", name: "my foo rule", type: "NODE", project_id: "project2", conditions: [{ operator: "MEMBER_OF", attribute: "CHEF_ORGS", values: ["Guardians of the Galaxy", "The Avengers"]}]}')" | jq .
```

#### Apply Project Rules

```
# curl -kH "api-token: $TOK" https://localhost/apis/iam/v2beta/apply-rules-start
```

Can look directly in elasticsearch for project tags (note that it's paginated though, so if you don't see any that might be why):

```
# curl $ELASTICSEARCH_URL/node-state/_search | jq . | grep -C 3 projects
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  221k  100  221k    0     0  12.0M      0 --:--:-- --:--:-- --:--:-- 12.0M
        "_source": {
          "chef_version": "12.21.3",
          "ipaddress": "227.238.101.188",
          "projects": [],
          "policy_revision": "",
          "roles": [],
          "lastCCRReceived": "2017-11-01T21:06:12Z",
--
...
--
        "_source": {
          "chef_version": "12.21.3",
          "ipaddress": "172.31.11.220",
          "projects": [
            "project1"
          ],
          "policy_revision": "",
```

And now we can see that our nodes moved after reindex:

```
# curl -H "projects: project1" -sH "api-token: $TOK" https://localhost/api/v0/cfgmgmt/nodes?pagination.size=1000 --insecure | jq . | grep "The Misfits" | wc -l
9
# curl -H "projects: (unassinged)" -sH "api-token: $TOK" https://localhost/api/v0/cfgmgmt/nodes?pagination.size=1000 --insecure | jq . | grep "The Misfits" | wc -l
0
```

#### Add a lot more nodes

```
# chef_load_nodes 10000
```

New nodes automagically get indexed:

```
# curl -H "projects: project1" -sH "api-token: $TOK" https://localhost/api/v0/cfgmgmt/nodes?pagination.size=10000 --insecure | jq . | grep "The Misfits" | wc -l
103
```

And now that we have enough, we can see the status and try out cancel:

```
# curl -sSkH "api-token: $TOK" https://a2-dev.test/apis/iam/v2beta/apply-rules-start && grpcurl --insecure -cert /hab/svc/config-mgmt-service/config/service.crt -key /hab/svc/config-mgmt-service/config/service.key localhost:10130 chef.automate.domain.authz.v2.Projects/ApplyRulesStatus
{}{
  "state": "running",
  "estimatedTimeComplete": "1970-01-01T00:00:00Z"
}

 # curl -sSkH "api-token: $TOK" https://a2-dev.test/apis/iam/v2beta/apply-rules-start && curl -sSkH "api-token: $TOK" https://a2-dev.test/apis/iam/v2beta/apply-rules-cancel

[213][default:/src:0]# grpcurl --insecure -cert /hab/svc/config-mgmt-service/config/service.crt -key /hab/svc/config-mgmt-service/config/service.key localhost:10130 chef.automate.domain.authz.v2.Projects/ApplyRulesStatus
{
  "state": "not_running",
  "estimatedTimeComplete": "0001-01-01T00:00:00Z",
  "percentageComplete": 1
}
```

Try it yourself: Query some `project2` nodes!

### :athletic_shoe: Demo Script / Repro Steps

rebuild components/automate-gateway
rebuild components/authz-service

Curl as above...

### :white_check_mark: Checklist

- [ ] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [x] Code actually executed?
- [x] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/authz-service/server/v2/project.go [hunk 1]
```diff
@@ -155,12 +155,6 @@ func (s *state) UpdateProject(ctx context.Context,
 			"error converting project with ID %q: %s", resp.ID, err.Error())
 	}
 
-	err = s.projectUpdateManager.Start()
-	if err != nil {
-		return nil, status.Errorf(codes.Internal,
-			"error starting project update %q: %s", resp.ID, err.Error())
-	}
-
```
Comment (msorens) [inline components/authz-service/server/v2/project.go]: Remove the POC call from UpdateProject.

Comment (tylercloke) [reply components/authz-service/server/v2/project.go]: Deleted unit test related to this code.

File: api/interservice/authz/v2/project.proto [hunk 2]
```diff
@@ -76,6 +76,7 @@ service Projects {
     rpc ListRulesForProject(ListRulesForProjectReq) returns (ListRulesForProjectResp) {};
     rpc DeleteRule(DeleteRuleReq) returns (DeleteRuleResp) {};
     rpc ListRulesForAllProjects(ListRulesForAllProjectsReq) returns (ListRulesForAllProjectsResp) {};
+    rpc ApplyRules(ApplyRulesReq) returns (ApplyRulesResp) {};
```
Comment (tylercloke) [inline api/interservice/authz/v2/project.proto]: Hmmm, on second though, should this go up in line 69 and be called `ProjectUpdateApply`? Seems like it goes with those APIs.

File: components/authz-service/server/v2/project.go [hunk 3]
```diff
@@ -155,23 +155,37 @@ func (s *state) UpdateProject(ctx context.Context,
 			"error converting project with ID %q: %s", resp.ID, err.Error())
 	}
 
-	err = s.projectUpdateManager.Start()
+	return &api.UpdateProjectResp{Project: apiProject}, nil
+}
+
+func (s *state) ApplyRulesStart(ctx context.Context, req *api.ApplyRulesStartReq) (*api.ApplyRulesStartResp, error) {
+	log.Info("apply project rules: START")
```
Comment (srenatus) [inline components/authz-service/server/v2/project.go]: ❓ `log`? Looks like this was **both** using the [global logrus logger](https://github.com/chef/automate/blob/fb00988690ad8fa4c56d661abb2f20a35c4abd4f/components/authz-service/server/v2/project.go#L18) and having [a logger in its state](https://github.com/chef/automate/blob/fb00988690ad8fa4c56d661abb2f20a35c4abd4f/components/authz-service/server/v2/project.go#L82). Let's avoid that, please 😉 
```suggestion
	s.log.Info("apply project rules: START")
```
same goes for the other `log.Info` calls below.

On that note, if we want every method to spit its name into our logs when called, we should do this via a grpc middleware. The gateway has a useful example wired up, [see here](https://github.com/chef/automate/blob/c32c9b7f5bdcd923008504f36a75456786ee9b0f/components/automate-gateway/gateway/server.go#L265-L287).

File: components/authz-service/server/v2/project.go [hunk 4]
```diff
@@ -155,23 +155,37 @@ func (s *state) UpdateProject(ctx context.Context,
 			"error converting project with ID %q: %s", resp.ID, err.Error())
 	}
 
-	err = s.projectUpdateManager.Start()
+	return &api.UpdateProjectResp{Project: apiProject}, nil
+}
+
+func (s *state) ApplyRulesStart(ctx context.Context, req *api.ApplyRulesStartReq) (*api.ApplyRulesStartResp, error) {
```
Comment (srenatus) [inline components/authz-service/server/v2/project.go]: This doesn't use `req`, let's replace it, **and more importantly** it doesn't use `ctx`. Is this something to fix in the `projectUpdateManager` API?
```suggestion
func (s *state) ApplyRulesStart(context.Context, *api.ApplyRulesStartReq) (*api.ApplyRulesStartResp, error) {
```

Looks like the same holds for the other methods, re: `ctx` not being used.

**Pull request merged**